src/assets/scripts/client/aircraft/FlightManagementSystem/Fms.js

```diff
@@ -0,0 +1,180 @@
+         * @type {array}
+         * @default []
+         */
+        this.legCollection = [];
+
+        /**
+         *
+         *
+         * @property category
+         * @type {string}
+         * @default ''
+         */
+        this.category = '';
+
+        this.init(aircraftInitProps);
+    }
+
+    /**
+     *
+     *
+     * @property currentWaypoint
+     * @return {FixModel|StandardRouteWaypointModel|undefined}
+     */
+    get currentWaypoint() {
+        return this.legCollection[0].currentWaypoint;
+    }
+
+    /**
+     *
+     *
+     * @property currentLeg
+     * @return {LegModel}
+     */
+    get currentLeg() {
+        return this.legCollection[0];
+    }
+
+    /**
+     *
+     *
+     * @for Fms
+     * @method init
+     * @param aircraftInitProps {object}
+     */
+    init(aircraftInitProps) {
+        this.category = aircraftInitProps.category;
+
+        this.legCollection = this._buildInitialLegCollection(aircraftInitProps);
+    }
+
+    /**
+     *
+     *
+     * @for LegModel
+     * @method destroy
+     */
+    destroy() {
+        this._navigationLibrary = null;
+        this._runway = '';
+        this.legCollection = [];
+        this.category = '';
+    }
+
+    /**
+     * Add a new `LegModel` to the left side of the `legCollection`
+     *
+     * @for Fms
+     * @method addLegToBeginning
+     * @param routeString
+     */
+    addLegToBeginning(routeString) {
+        const legModel = new LegModel(routeString, this._runway, this.category, this._navigationLibrary);
+
+        this.legCollection.unshift(legModel);
+    }
+
+    /**
+     * Move to the next possible waypoint
+     *
+     * This could be the next waypoint in the current leg,
+     * or the first waypoint in the next leg.
+     *
+     * @for LegModel
+     * @method nextWaypoint
+     */
+    nextWaypoint() {
+        if (!this.currentLeg.hasNextWaypoint()) {
+            this._moveToNextLeg();
+        }
+
+        this._moveToNextWaypointInLeg();
+    }
+
+    /**
+     *
+     *
+     * @for LegModel
+     * @method _buildInitialLegCollection
+     * @param aircraftInitProps {object}
+     * @private
+     */
+    _buildInitialLegCollection(aircraftInitProps) {
+        const { route } = aircraftInitProps;
+        const routeStringSegments = routeStringFormatHelper(route);
+        const legsForRoute = _map(routeStringSegments, (routeSegment) => {
+            return new LegModel(routeSegment, this._runway, this.category, this._navigationLibrary);
+        });
+
+        return legsForRoute;
+    }
+
+    /**
+     *
+     *
+     * @for Fms
+     * @method _moveToNextWaypointInLeg
+     * @private
+     */
+    _moveToNextWaypointInLeg() {
+        this.currentLeg.moveToNextWaypoint();
+    }
+
+    /**
+     *
+     * @for Fms
+     * @method _moveToNextLeg
+     */
+    _moveToNextLeg() {
+        this.currentLeg.destroy();
+        // this is mutable
+        this.legCollection.shift();
```

Comment:
This idea of removing past navigation information is fine, but we do need to store somewhere the full route, for purposes such as `runSayRoute` or displaying the flightplan route in a flight strip. This could be a part of the aircraft instead of the fms if desired.